File: Leetcode_POTD_2025/7_JULY/3307_Find_the_K_th_Character_in_String_Game_II.py
class Solution(object):
    def kthCharacter(self, k, operations): 
        prime_val = 'a'
        phase_shift = 0
        current_span = 1
        
        for _ in operations: 
            current_span <<= 1

        for morph_type in reversed(operations): 
            current_span >>= 1
            if k > current_span: 
                k -= current_span 
                if morph_type == 1:
                    phase_shift += 1
        
        final_char_code = ord(prime_val) + phase_shift
        return chr(ord('a') + (final_char_code - ord('a')) % 26)

File: Leetcode_POTD_2025/7_JULY/3307_Find_the_K_th_Character_in_String_Game_II.java
class Solution {
    public char kthCharacter(long k, int[] operations) {
        char prime_val = 'a';
        int phase_shift = 0;
        long current_span = 1;
        
        for (int op : operations) {
            current_span <<= 1;
        }

        for (int i = operations.length - 1; i >= 0; i--) {
            int morph_type = operations[i];
            current_span >>= 1;
            if (k > current_span) {
                k -= current_span;
                if (morph_type == 1) {
                    phase_shift++;
                }
            }
        }
        
        int final_char_code = prime_val + phase_shift;
        return (char) ('a' + (final_char_code - 'a') % 26);
    }
}
